src/Libraries/Nop.Data/Migrations/UpgradeTo460/DataMigration.cs

```diff
@@ -0,0 +1,104 @@
+﻿using FluentMigrator;
+using Nop.Core.Domain.Customers;
+using Nop.Data.Mapping;
+using Nop.Data.Extensions;
+using Nop.Core.Domain.Directory;
+using Nop.Core.Domain.Localization;
+using Nop.Core.Domain.Tax;
+using System.Data;
+using System;
+using System.Linq;
+using Nop.Core.Domain.Common;
+
+namespace Nop.Data.Migrations.UpgradeTo460
+{
+    [NopMigration("2022-02-03 00:00:00", "4.60.0", UpdateMigrationType.Data, MigrationProcessType.Update)]
+    public class DataMigration : Migration
+    {
+        private readonly INopDataProvider _dataProvider;
+
+        public DataMigration(INopDataProvider dataProvider)
+        {
+            _dataProvider = dataProvider;
+        }
+
+        /// <summary>
+        /// Collect the UP migration expressions
+        /// </summary>
+        public override void Up()
+        {
+            MigrateCustomerAttributes();
+        }
+
+        public override void Down()
+        {
+            //add the downgrade logic if necessary 
+        }
+
+        private void MigrateCustomerAttributes()
+        {
+            // customer attibute values to customer table column values
+            var customerRole = _dataProvider.GetTable<CustomerRole>().FirstOrDefault(cr => cr.SystemName == NopCustomerDefaults.RegisteredRoleName);
+            var customerRoleId = customerRole?.Id ?? 0;
+
+            var query = from c in _dataProvider.GetTable<Customer>()
+                        join crm in _dataProvider.GetTable<CustomerCustomerRoleMapping>() on c.Id equals crm.CustomerId
+                        where !c.Deleted && (customerRoleId == 0 || crm.CustomerRoleId == customerRoleId)
+                        select new { Customer = c, CustomerRoleMapping = crm };
+
+            var pageIndex = 0;
+            var pageSize = 500;
+
+            while (true)
+            {
+                var customers = query.Select(x => x.Customer).Skip(pageIndex * pageSize).Take(pageSize).ToList();
+                if (!customers.Any())
+                    break;
+
+                for (var i = 0; i < customers.Count; i++)
+                {
+                    var customer = customers[i];
+                    var customerAttributes = _dataProvider.GetTable<GenericAttribute>()
+                        .Where(ga => ga.KeyGroup == nameof(Customer) && ga.EntityId == customer.Id).ToList();
+
+                    if (!customerAttributes.Any())
+                        continue;
+
+                    customer.FirstName = customerAttributes.FirstOrDefault(ga => ga.Key == "FirstName")?.Value;
```

Comment:
Let's use `nameof(Customer.FirstName)` instead of string values, it will be more reliable. Same for all other properties. Do the names of the properties completely match with the names of the attributes that were before?